### PR TITLE
Source section less node

### PR DESCRIPTION
**Pull request by jtulach**
According to discussion in #25 it seems that there is a general consensus that SourceSections should be handled by individual languages and the interface in Truffle API simplified to:

- Node.getSourceSection() that can be implemented by subclasses
- RootNode(SourceSection, ...) constructor accepting SourceSection as majority of RootNode instances have source section

The other methods are deprecated while the functionality is kept. Once the deprecated methods are removed, the Node#sourceSection field will be removed as well. RootNode will then get such field to hold the value provided in constructor.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,2 +1,12 @@
...
+to your first pull request indicating the name you used to sign the OCA if it isn't clear
+from your github profile.
+
+Pull requests can only be merged by committers and a committer cannot merge
+his or her own pull requests.
+
+When creating a pull request, make sure you designate an assignee to ensure
+the request is processed in a timely manner.
+
+[1]: http://www.oracle.com/technetwork/community/oca-486395.html
```
Comment (smarr) [inline CONTRIBUTING.md]: This looks like it should not be part of this pull request?

File: truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java [hunk 2]
```diff
@@ -119,14 +127,31 @@ public NodeCost getCost() {
...
+     *
+     * {@codesnippet source.section.fixed}
+     *
+     * To create a node which can associate and change the {@link SourceSection} later at any point
+     * of time use:
+     *
+     * {@codesnippet source.section.mutable}
      *
      * @return the source code represented by this Node
      */
```
Comment (smarr) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: Does this comment need a transitory note? Mostly, because the implementation does not yet return null?

Comment (jtulach) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: If you use just non-deprecated API it returns null, so I think updating the Javadoc "in advance" is proper thing to do.

----- Reply message -----
Od: "Stefan Marr" <notifications@github.com>
Komu: "graalvm/truffle" <truffle@noreply.github.com>
Kopie: "Jaroslav Tulach" <jaroslav.tulach@oracle.com>
Předmět: [truffle] Source section less node (#35)
Datum: čt, úno 4, 2016 12:06

In truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java:
>      public void clearSourceSection() {
>          this.sourceSection = null;
>      }
>  
>      /**
> -     * Retrieves the segment of guest language source code that is represented by this Node.
> +     * Retrieves the segment of guest language source code that is represented by this Node. The
> +     * default implementation of this method returns <code>null</code>. If your node represents a
> +     * segment of the source code, override this method and return a <code>final</code> or
> +     * {@link CompilationFinal} field in your node to the caller.
> +     *
> +     * To define node with <em>fixed</em> {@link SourceSection} that doesn't change after node
> +     * construction use:
> +     *
> +     * {@codesnippet source.section.fixed}
> +     *
> +     * To create a node which can associate and change the {@link SourceSection} later at any point
> +     * of time use:
> +     *
> +     * {@codesnippet source.section.mutable}
>       *
>       * @return the source code represented by this Node
>       */

Does this comment need a transitory note? Mostly, because the implementation does not yet return null?

&mdash;

File: truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java [hunk 3]
```diff
@@ -119,14 +127,31 @@ public NodeCost getCost() {
...
+     * To define node with <em>fixed</em> {@link SourceSection} that doesn't change after node
+     * construction use:
+     *
+     * {@codesnippet source.section.fixed}
+     *
+     * To create a node which can associate and change the {@link SourceSection} later at any point
+     * of time use:
+     *
+     * {@codesnippet source.section.mutable}
      *
```
Comment (chumer) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: I am still not convinced by the usefulness of these codesnippets. I would even say that most GL devs don't read javadocs but are browsing using the IDE where codesnippets dont translate nicely.

Comment (smarr) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: Hm, well, the issue would be little less problematic if at least the names would be proper JavaDoc names (like for `@link`) that one can click on to navigate to the example.

Comment (woess) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: yes, including at least the classname would help a bit to find the snippet within an IDE...

Comment (jtulach) [inline truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java]: I would prefer to leave the snippet discussion in #19 - I understand they are source of tension. That said the idea to make them behave like a link or at least include the name of source file is certainly worth exploring.

**Comment by woess** [comment]
@jtulach I'd like to hear other people's opinion about this as well, but I think getSourceSection() should stay public.

**Comment by mlvdv** [comment]
`getSourceSection()` should be public.  There will be tools (none yet) for which precise source attribution of nodes will be a requirement.  This translates to the the invariant that a node was produced by parsing the text represented by its `SourceSection`.  In contrast, `getEncapsulatingSourceSection()` should mean something else:  what is the nearest enclosing syntactic unit that this `Node` is part of?  
[Aside: `getEncapsulatingSourceSection()`  only makes sense reliably when ASTs are "well-behaved" with respect to source attribution, something discussed in some detail elsewhere.  Well-behaved ASTs are important for some kinds of tools, but this imposes a constraint on AST design that I've often seen violated, e.g. in *javac*.]
@mickjordan may want to comment; we spoke about this while the FastR AST was under development.

**Comment by jtulach** [comment]
I see. A real use-case for a tool that we yet have to imagine.

If I accept as given that getSourceSection is needed - do we need getEncapsulatingSourceSection() helper method at all? It seems to me like utility method that iterates over chain of nodes. Utility methods don't belong into core API. Each tool write the loop on their own.

**Comment by mickjordan** [comment]
On 2/4/16 11:38 AM, Michael Van De Vanter wrote:
>
> |getSourceSection()| should be public. There will be tools (none yet) 
> for which precise source attribution of nodes will be a requirement. 
> This translates to the the invariant that a node was produced by 
> parsing the text represented by its |SourceSection|. In contrast, 
> |getEncapsulatingSourceSection()| should mean something else: what is 
> the nearest enclosing syntactic unit that this |Node| is part of?
>
> [Aside: |getEncapsulatingSourceSection()| only makes sense reliably 
> when ASTs are "well-behaved" with respect to source attribution, 
> something discussed in some detail elsewhere. Well-behaved ASTs are 
> important for some kinds of tools, but this imposes a constraint on 
> AST design that I've often seen violated, e.g. in /javac/.]
> @mickjordan <https://github.com/mickjordan> may want to comment; we 
> spoke about this while the FastR AST was under development.
>
>
In FastR we removed all uses of getEncapsulatingSourceSection and throw 
an exception if anyone uses it. Our model is that you first find the 
RSyntaxNode (getRSyntaxNode) and then call getSourceSection, which must 
not fail. We actually try hard to avoid accessing the source section 
until we really need it, like printing an error with context. We do 
support a hierarchical search by default for getRSyntaxNode, but a node 
can override if it's node structure is not hierarchical. As you recall I 
tried to get this model into truffle without success, so we have a 
special subclass of Node that contains the new methods and overrides of 
the Truffle ones. It's a slight pain to have to remember to use this in 
preference to Node. You can subclass Node directly but you have to 
follow some rules to be compliant with the model.

Mick

**Comment by jtulach** [comment]
I see that FastR [overrides getEncapsulatingSourceSetion](http://lahoda.info/index/ui/index.html#/showCode?path=graal&relative=fastr%2Fcom.oracle.truffle.r.runtime%2Fsrc%2Fcom%2Foracle%2Ftruffle%2Fr%2Fruntime%2Fnodes%2FRBaseNode.java&goto=METHOD:com.oracle.truffle.r.runtime.nodes.RBaseNode:getEncapsulatingSourceSection:%28%29Lcom%2Foracle%2Ftruffle%2Fapi%2Fsource%2FSourceSection%3B) to delegate to **protected** method that finds the right node and extracts the section from it. I see [four overrides of the method](http://lahoda.info/index/ui/index.html#/usages?signature=METHOD:com.oracle.truffle.r.runtime.nodes.RBaseNode:getRSyntaxNode:%28%29Lcom%2Foracle%2Ftruffle%2Fr%2Fruntime%2Fnodes%2FRSyntaxNode%3B). FastR overrides *getSourceSection*, but only to throw exceptions in unexpected cases. FastR calls getSourceSection - but mostly only because it passes the section somewhere else (*assignSourceSection* or constructor of other node).

As far as I can tell FastR would benefit from language controlled *find/getSourceSection* for a node.

**Comment by chumer** [comment]
I like the change. +1

I'd leave getSourceSection public for now, but maybe revisit this in another changeset? I think the instrumentation tools could use accessors to access the source section. Tools can access source sections using the EventContext via public API.

@chrisseaton @woess please also accept or deny this change 

@jtulach After that please rebase and merge.

File: truffle/com.oracle.truffle.api.dsl/src/com/oracle/truffle/api/dsl/internal/DSLShare.java [hunk 4]
```diff
@@ -146,7 +146,8 @@ public T call() {
 
     private static void updateSourceSection(Node oldNode, Node newNode) {
         if (newNode.getSourceSection() == null) {
-            newNode.assignSourceSection(oldNode.getSourceSection());
+            throw new IllegalArgumentException("Can't update source: " + oldNode);
+            // newNode.assignSourceSection(oldNode.getSourceSection());
```
Comment (chrisseaton) [inline truffle/com.oracle.truffle.api.dsl/src/com/oracle/truffle/api/dsl/internal/DSLShare.java]: Comment left in by mistake?

File: truffle/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/SourceSectionTest.java [hunk 5]
```diff
@@ -108,19 +121,28 @@ Object do4(Object a) {
...
+        TestRootNode<NodeWithFixedSourceSection> root = createRootPrefix(SourceSectionTestFactory.NodeWithFixedSourceSectionFactory.getInstance(), true, section);
         expectSourceSection(root.getNode(), section);
         assertThat((int) executeWith(root, 1), is(1));
         expectSourceSection(root.getNode(), section);
     }
 
     @NodeChild("a")
-    static class SourceSection1 extends ValueNode {
+    static class NodeWithFixedSourceSection extends ValueNode {
+        // BEGIN: NodeWithFixedSourceSection
```
Comment (chrisseaton) [inline truffle/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/SourceSectionTest.java]: Are these comments directives to some tool? What do they do?

**Comment by chrisseaton** [comment]
I'm not sure what the final proposed change is, as @chumer's last comments suggests more changes which may or may not be accepted, so it's hard to accept right now.

I currently use `getEncapsulatingSourceSection` on arbitrary foreign nodes to print backtraces for other language's stack frames in Ruby, so I'd need that functionality to remain or an alternative.

I don't mind maintaining the `SourceSection` field myself in Ruby, that seems like a good idea so languages can do it in different ways, but I think `getSourceSection` (maybe with the semantics of `getEncapsulatingSourceSection`) should stay, given the example of my backtraces.

File: truffle/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/SourceSectionTest.java (hunk 5 above)
Comment (chumer) [inline truffle/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/SourceSectionTest.java]: See https://github.com/graalvm/truffle/issues/19

File: CONTRIBUTING.md (hunk 1 above)
Comment (woess) [inline CONTRIBUTING.md]: yes, please rebase.

**Comment by woess** [comment]
@chumer @jtulach 
I accept the change (needs to be rebased though).
For now, getSourceSection(), getEncapsulatingSourceSection(), and RootNode should be kept as is, and any suggestion to change them is out of scope of this PR.

File: truffle/com.oracle.truffle.api.dsl/src/com/oracle/truffle/api/dsl/internal/DSLShare.java (hunk 4 above)
Comment (jtulach) [inline truffle/com.oracle.truffle.api.dsl/src/com/oracle/truffle/api/dsl/internal/DSLShare.java]: I kept the comment as I expected the exception to be raised at some places, but it wasn't raised yet. But we haven't yet executed all our language tests. Maybe it is wiser to keep it for now.

**Comment by jtulach** [comment]
OK, this change request will just deprecate Node(SourceSection) constructor and Node#assignSourceSection and Node#cleanSourceSection. Unless there are objections, let's merge tomorrow.

**Comment by jtulach** [comment]
No objections. Merging.
